concourse/pipelines/templates/build_pipeline-tpl.yml

```
@@ -1,1842 +1,326 @@
-    username: _json_key
-    password: ((ud/pxf/secrets/pxf-cloudbuild-service-account-key))
+    [[macros.resource_registry_image('ccp-7-image', 'gcr.io/data-gpdb-public-images/ccp', 'latest')]]
 
-{% set gp_ver = 6 %}
-- name: gpdb[[gp_ver]]-pxf-dev-ubuntu18-image
-  type: registry-image
-  icon: docker
-  source:
-    repository: gcr.io/data-gpdb-ud/gpdb-pxf-dev/gpdb[[gp_ver]]-ubuntu18.04-test-pxf
-    tag: latest
-    username: _json_key
-    password: ((ud/pxf/secrets/pxf-cloudbuild-service-account-key))
-
-{% set gp_ver = None %}
-
-- name: rpmrebuild-centos7-image
-  type: registry-image
-  icon: docker
-  source:
-    repository: gcr.io/data-gpdb-ud/rpmrebuild-centos7
-    tag: latest
-    username: _json_key
-    password: ((ud/pxf/secrets/pxf-cloudbuild-service-account-key))
-
-- name: rpmrebuild-rocky8-image
-  type: registry-image
-  icon: docker
-  source:
-    repository: gcr.io/data-gpdb-ud/rpmrebuild-rocky8
-    tag: latest
-    username: _json_key
-    password: ((ud/pxf/secrets/pxf-cloudbuild-service-account-key))
-
-- name: ccp-7-image
-  type: registry-image
-  icon: docker
-  source:
-    repository: gcr.io/data-gpdb-public-images/ccp
-    tag: latest
-
-- name: google-cloud-sdk-slim-image
-  type: registry-image
-  icon: docker
-  source:
-    repository: google/cloud-sdk
-    tag: 402.0.0-slim
+    [[macros.resource_registry_image('google-cloud-sdk-slim-image', 'google/cloud-sdk', '402.0.0-slim')]]
 
+## ---------- Slack Notification ----------
 {% if slack_notification %}
-- name: slack-alert
-  type: slack-notification
-  source:
-    url: ((ud/pxf/secrets/ud-pipeline-bot-ud-dev-webhook))
-
-- name: notify-releng
-  type: slack-notification
-  source:
-    url: ((ud/pxf/secrets/ud-pipeline-bot-gp-releng-webhook))
+    [[macros.resource_slack_notification('slack-alert', 'ud-pipeline-bot-dev-webhook')]]
+    [[macros.resource_slack_notification('notify-releng', 'ud-pipeline-bot-gp-releng-webhook')]]
 {% endif %}
 
 ## ---------- Greenplum Packages ----------
-{% set gp_ver = 5 %}
-{% for i in range(num_gpdb5_versions) %}
-- name: gpdb[[gp_ver]]_rhel7_rpm_latest-[[i]]
-  type: gcs
-  icon: google-drive
-  source:
-    bucket: ((ud/pxf/common/pivnet-artifacts-bucket-name))
-    json_key: ((ud/pxf/secrets/pxf-storage-service-account-key))
-    regexp: latest-[[i]]_gpdb[[gp_ver]]/greenplum-db-([[gp_ver]].*)-rhel7-x86_64.rpm
-{% endfor %}{# range(num_gpdb5_versions) #}
-{% set gp_ver = None %}
-
-{% set gp_ver = 6 %}
-{% for i in range(num_gpdb6_versions) %}
-- name: gpdb[[gp_ver]]_rhel7_rpm_latest-[[i]]
-  type: gcs
-  icon: google-drive
-  source:
-    bucket: ((ud/pxf/common/pivnet-artifacts-bucket-name))
-    json_key: ((ud/pxf/secrets/pxf-storage-service-account-key))
-    regexp: latest-[[i]]_gpdb[[gp_ver]]/greenplum-db-([[gp_ver]].*)-rhel7-x86_64.rpm
+{% call(x) macros.for_each_config(build_test_pxf_combinations) %}
+    [[macros.set_num_versions(x, num_gpdb5_versions, num_gpdb6_versions, num_gpdb7_versions)]]
```

it feels like a dictionary would be better here:
```suggestion
    [[macros.set_num_versions(x, gp_num_versions)]]
```
where gp_num_versions = { '5': num_gpdb5_versions, '6': num_gpdb6_versions, '7': num_gpdb7_versions}